sklearn/utils/fixes.py

```diff
@@ -106,6 +108,39 @@ def _eigh(*args, **kwargs):
         return scipy.linalg.eigh(*args, eigvals=eigvals, **kwargs)
 
 
+def _with_config(delayed_func, config):
+    """Helper function that intends to attach a config to a delayed function."""
+    if hasattr(delayed_func, "with_config"):
+        return delayed_func.with_config(config)
+    else:
+        warnings.warn(
+            "You are using `sklearn.utils.fixes.Parallel` that intend to attached a "
+            "configuration to a delayed function. However, the function used for "
+            "delaying the function does not expose `with_config`. Use "
+            "`sklearn.utils.fixes.delayed` for this purpose. A default configuration "
+            "is used instead.",
+            UserWarning,
+        )
+        return delayed_func
+
+
+class Parallel(joblib.Parallel):
```

Comment:
If we want to support `fixes.delayed` and `fixes.Parallel` in the long term, then I think they need to be in another module, like `utils.parallel`. Introducing a `utils.fixes.Parallel` would mean a deprecation cycle while we move to `utils.parallel`.

For the `1.2.1` release, we can have a private `utils._parallel` submodule that includes `delayed` and `Parallel` and in `1.3` we make `utils.parallel` public. What do you think?

Reply:
I agree that we can move those to a new module. If we come to deprecate those tools, we can always deprecate the entire module.

Regarding having a private module in `1.2.1` becoming public in `1.3`, this is a bit complex. We will start issuing warnings asking to use those tools instead of joblib.
Therefore, we should not issue warnings mentioning an import from a private module that will change in the next version.

I understand the issue of proposing an API change in a bug fix release but I don't think that we will have a choice here if we want people to use it in third-party library.

@thomasjpfan do you think of an alternative that I did not think of?

Reply:
> I understand the issue of proposing an API change in a bug fix release but I don't think that we will have a choice here if we want people to use it in third-party library.

I can not think of anything else. I'm okay with introducing a new `utils.parallel` submodule in 1.2.1. Given that this is a developer API, it should not have too much impact on users and greatly benefit third party developers.